lib/modules/manager/composer/schema.ts

```diff
@@ -0,0 +1,313 @@
+  url: z.string(),
+});
+export type GitRepo = z.infer<typeof GitRepo>;
+
+export const PathRepo = z.object({
+  type: z.literal('path'),
+  url: z.string(),
+});
+export type PathRepo = z.infer<typeof PathRepo>;
+
+export const Repo = z.discriminatedUnion('type', [
+  ComposerRepo,
+  GitRepo,
+  PathRepo,
+]);
+export type Repo = z.infer<typeof ComposerRepo>;
+
+export const NamedRepo = z.discriminatedUnion('type', [
+  ComposerRepo,
+  GitRepo.extend({ name: z.string() }),
+  PathRepo.extend({ name: z.string() }),
+]);
+export type NamedRepo = z.infer<typeof NamedRepo>;
+
+const DisablePackagist = z.object({ type: z.literal('disable-packagist') });
+export type DisablePackagist = z.infer<typeof DisablePackagist>;
+
+export const ReposRecord = LooseRecord(z.union([Repo, z.literal(false)]), {
+  onError: ({ error: err }) => {
+    logger.warn({ err }, 'Composer: error parsing repositories object');
+  },
+}).transform((repos) => {
+  const result: (NamedRepo | DisablePackagist)[] = [];
+  for (const [name, repo] of Object.entries(repos)) {
+    if (repo === false) {
+      if (name === 'packagist' || name === 'packagist.org') {
+        result.push({ type: 'disable-packagist' });
+      }
+      continue;
+    }
+
+    if (repo.type === 'path' || repo.type === 'git') {
+      result.push({ name, ...repo });
+      continue;
+    }
+
+    if (repo.type === 'composer') {
+      result.push(repo);
+      continue;
+    }
+  }
+
+  return result;
+});
+export type ReposRecord = z.infer<typeof ReposRecord>;
+
+export const ReposArray = LooseArray(
+  z.union([
+    NamedRepo,
+    z
+      .union([
+        z.object({ packagist: z.literal(false) }),
+        z.object({ 'packagist.org': z.literal(false) }),
+      ])
+      .transform((): DisablePackagist => ({ type: 'disable-packagist' })),
+  ]),
+  {
+    onError: ({ error: err }) => {
+      logger.warn({ err }, 'Composer: error parsing repositories array');
+    },
+  }
+).transform((repos) => repos.filter((x): x is NamedRepo => x !== null));
+export type ReposArray = z.infer<typeof ReposArray>;
+
+export const Repos = z
+  .union([ReposRecord, ReposArray])
+  .default([]) // Prevents warnings for packages without repositories field
+  .catch(({ error: err }) => {
+    logger.warn({ err }, 'Composer: repositories parsing error');
+    return [];
+  })
+  .transform((repos) => {
+    let packagist = true;
+    const registryUrls: string[] = [];
+    const gitRepos: Record<string, GitRepo> = {};
+    const pathRepos: Record<string, PathRepo> = {};
+
+    for (const repo of repos) {
+      if (repo.type === 'composer') {
+        registryUrls.push(repo.url);
+      } else if (repo.type === 'git') {
+        gitRepos[repo.name] = repo;
+      } else if (repo.type === 'path') {
+        pathRepos[repo.name] = repo;
+      } else if (repo.type === 'disable-packagist') {
+        packagist = false;
+      }
+    }
+
+    if (packagist) {
+      registryUrls.push('https://packagist.org');
```

Comment:
```suggestion
      registryUrls.unshift('https://packagist.org');
```
so it's first instead of last checked? i think most packages are on that registry.